src/main/java/de/tum/in/www1/artemis/service/competency/CompetencyProgressService.java

```diff
@@ -217,26 +204,187 @@ public CompetencyProgress updateCompetencyProgress(Long competencyId, User user)
+
+        double achievedPointsInDifficulty = participantScores.stream().filter(info -> info.difficulty() == difficultyLevel)
+                .mapToDouble(CompetencyExerciseMasteryCalculationDTO::lastPoints).sum();
+        double pointsInCompetencyInDifficulty = exerciseInfos.stream().filter(info -> info.difficulty() == difficultyLevel)
+                .mapToDouble(CompetencyExerciseMasteryCalculationDTO::maxPoints).sum();
+
+        double proportionOfAchievedPointsInDifficulty = achievedPointsInDifficulty / achievedPoints;
+        double proportionOfPointsInCompetencyInDifficulty = pointsInCompetencyInDifficulty / pointsInCompetency;
+
+        return proportionOfAchievedPointsInDifficulty - proportionOfPointsInCompetencyInDifficulty;
+    }
+
+    /**
+     * Calculate the quick solve confidence heuristic for the given user in a competency based on the exercises linked to the competency.
+     * If the user has solved a high proportion of programming exercises with a high score in a short amount of time, the confidence should be higher.
+     *
+     * @param participantScores the participant scores for the exercises linked to the competency
+     * @return The quick solve confidence heuristic
+     */
+    private double calculateQuickSolveConfidenceHeuristic(@NotNull Set<CompetencyExerciseMasteryCalculationDTO> participantScores) {
+        Set<CompetencyExerciseMasteryCalculationDTO> programmingParticipationScores = participantScores.stream()
+                .filter(CompetencyExerciseMasteryCalculationDTO::isProgrammingExercise).collect(Collectors.toSet());
+
+        if (programmingParticipationScores.isEmpty()) {
+            return 0;
+        }
+
+        double numberOfQuicklySolvedProgrammingExercises = programmingParticipationScores.stream()
+                .filter(info -> info.lastScore() >= MIN_SCORE_GREEN && info.submissionCount() <= 3).count();
+
+        double quickSolveConfidence = numberOfQuicklySolvedProgrammingExercises / programmingParticipationScores.size();
+
+        return Math.clamp(quickSolveConfidence, -0.25, 0.25);
+    }
+
+    private void setConfidenceReason(CompetencyProgress competencyProgress, double recencyConfidence, double difficultyConfidence, double quickSolveConfidence) {
+        if (competencyProgress.getConfidence() < 0.95) {
+            double minConfidenceHeuristic = Math.min(recencyConfidence, difficultyConfidence);
+            if (recencyConfidence == minConfidenceHeuristic) {
+                competencyProgress.setConfidenceReason(CompetencyProgressConfidenceReason.RECENT_SCORES_LOWER);
+            }
+            else {
+                // quickSolveConfidence cannot be negative therefore we don't check it here
+                competencyProgress.setConfidenceReason(CompetencyProgressConfidenceReason.MORE_EASY_POINTS);
+            }
+        }
+        else if (competencyProgress.getConfidence() > 1.05) {
```

Comment:
What is the reasoning behind `0.95` and `1.05` could we add variables for the magic numbers and document in the javadoc why these values have been chosen (if there was a particular reason)?